From c9fb54f6728525f63324cb14293e9a34e52cf758 Mon Sep 17 00:00:00 2001
From: Rob Herring <r.herring@freescale.com>
Date: Thu, 31 Jul 2008 09:35:33 -0500
Subject: [PATCH] ENGR00086177 Enable IPU DI for MX31 cpu clock change

The IPU clock and DI module must be enabled on MX31 and MX32 or PDR0 updates
for DVFS will not take effect. There is no indication from the h/w other than
the frequencies will be wrong after a clock rate change.

Signed-off-by: Rob Herring <r.herring@freescale.com>
---
 arch/arm/mach-mx3/clock.c |   14 ++++++++++++++
 1 files changed, 14 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-mx3/clock.c b/arch/arm/mach-mx3/clock.c
index 630cadd..971754d 100644
--- a/arch/arm/mach-mx3/clock.c
+++ b/arch/arm/mach-mx3/clock.c
@@ -1269,6 +1269,8 @@ static int cpu_clk_set_wp(int wp)
 	u32 pdr0;
 	u32 vscnt = MXC_CCM_PMCR0_VSCNT_2;
 	u32 udsc = MXC_CCM_PMCR0_UDSC_DOWN;
+	u32 ipu_base = IO_ADDRESS(IPU_CTRL_BASE_ADDR);
+	u32 ipu_conf;
 
 	if (wp >= cpu_wp_nr || wp < 0) {
 		printk(KERN_ERR "Wrong wp: %d for cpu_clk_set_wp\n", wp);
@@ -1313,6 +1315,13 @@ static int cpu_clk_set_wp(int wp)
 
 	__raw_writel(pmcr0, MXC_CCM_PMCR0);
 
+	/* IPU and DI submodule must be on for PDR0 update to take effect */
+	if (!clk_get_usecount(&ipu_clk))
+		ipu_clk.enable(&ipu_clk);
+	ipu_conf = __raw_readl(ipu_base);
+	if (!(ipu_conf & 0x40))
+		__raw_writel(ipu_conf | 0x40, ipu_base);
+
 	__raw_writel((pdr0 & ~MXC_PDR0_MAX_MCU_MASK) | p->pdr0_reg,
 		     MXC_CCM_PDR0);
 
@@ -1334,6 +1343,11 @@ static int cpu_clk_set_wp(int wp)
 
 	cpu_curr_wp = wp;
 
+	/* Restore IPU_CONF setting */
+	__raw_writel(ipu_conf, ipu_base);
+	if (!clk_get_usecount(&ipu_clk))
+		ipu_clk.disable(&ipu_clk);
+
 	if (wp == cpu_wp_nr - 1) {
 		init_timer(&dptcen_timer);
 		dptcen_timer.expires = jiffies + 2;
-- 
1.5.4.4

